src/Renci.SshNet/Security/Cryptography/Ciphers/AesGcmCipher.cs

```diff
@@ -0,0 +1,137 @@
+                return 16;
+            }
+        }
+
+        /// <inheritdoc/>
+        public override int TagSize
+        {
+            get
+            {
+                return 16;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AesGcmCipher"/> class.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <param name="iv">The IV.</param>
+        public AesGcmCipher(byte[] key, byte[] iv)
+            : base(key)
+        {
+            _nonce = iv.Take(12);
+#if NET8_0_OR_GREATER
+            _aesGcm = new AesGcm(key, TagSize);
+#else
+            _aesGcm = new AesGcm(key);
+#endif
+        }
+
+        /// <summary>
+        /// Encrypts the specified input.
+        /// </summary>
+        /// <param name="input">The input.</param>
+        /// <param name="offset">The zero-based offset in <paramref name="input"/> at which to begin encrypting.</param>
+        /// <param name="length">The number of bytes to encrypt from <paramref name="input"/>.</param>
+        /// <returns>
+        /// The packet length field + cipher text + tag.
+        /// </returns>
+        public override byte[] Encrypt(byte[] input, int offset, int length)
+        {
+            // [outbound sequence field][packet length field][padding length field sz][payload][random paddings]
+            // [----4 bytes----(offset)][------4 bytes------][----------------Plain Text---------------(length)]
+            var packetLengthField = new ReadOnlySpan<byte>(input, offset, 4);
+            var plainText = new ReadOnlySpan<byte>(input, offset + 4, length - 4);
+
+            var output = new byte[length + TagSize];
+            packetLengthField.CopyTo(output);
+            var cipherText = new Span<byte>(output, 4, length - 4);
+            var tag = new Span<byte>(output, length, TagSize);
+
+            _aesGcm.Encrypt(_nonce, plainText, cipherText, tag, packetLengthField);
+
+            IncrementCounter();
+
+            return output;
+        }
+
+        /// <summary>
+        /// Decrypts the specified input.
+        /// </summary>
+        /// <param name="input">The input.</param>
+        /// <param name="offset">The zero-based offset in <paramref name="input"/> at which to begin decrypting and authenticating.</param>
+        /// <param name="length">The number of bytes to decrypt and authenticate from <paramref name="input"/>.</param>
+        /// <returns>
+        /// The packet length field + plain text.
+        /// </returns>
+        public override byte[] Decrypt(byte[] input, int offset, int length)
+        {
+            // [inbound sequence field][packet length field][padding length field sz][payload][random paddings][Authenticated TAG]
+            // [----4 bytes---(offset)][------4 bytes------][------------------Cipher Text--------------------][---TAG---(length)]
+            var packetLengthField = new ReadOnlySpan<byte>(input, offset, 4);
```

Comment:
Similarly, move the explanation to /// xmldoc

Reply:
Actually, I once put the xml doc comments. There are 2 small issues
1. I once put the structure in xml doc `remarks` section. However, it requires a period at the end which means I need add some paragraphs after `[---TAG---(length)]`
   See https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1629.md
3. xml doc comments are primarily for lib consumers. This kind of details might be more suitable for lib developers. Is there any existing simlar details in xml doc comments?

Reply:
> Actually, I once put the xml doc comments. There are 2 small issues
> 
> 1. I once put the structure in xml doc `remarks` section. However, it requires a period at the end which means I need add some paragraphs after `[---TAG---(length)]`
>    See https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1629.md
> 2. xml doc comments are primarily for lib consumers. This kind of details might be more suitable for lib developers. Is there any existing simlar details in xml doc comments?

You could try using the `<code>` element: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#code

Reply:
@jscarle Thanks. I did think to put it inside `<code>` element. However I doubted it is code. It is quite different with C#. I'm not native English speaker but just now I realized the word "code" can be very generic.

Reply:
Done